src/main/java/com/google/sps/servlets/StudyServlet.java

```diff
@@ -0,0 +1,200 @@
+    long time = DateTimeUtils.currentTimeMillis();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    try {
+      cardEntity = datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      cardEntity = null;
+    }
+
+    // Handles a null entity
+    if (cardEntity == null) {
+      return;
+    }
+    Card card = new Card(cardEntity);
+    if (answeredCorrectly) {
+      newScore = incFamiliarityScore(time, card.getFamiliarityScore(), card.getTimeTested());
+    } else {
+      newScore = decFamiliarityScore(time, card.getFamiliarityScore(), card.getTimeTested());
+    }
+    cardEntity.setProperty("familiarityScore", newScore);
+    cardEntity.setProperty("timeTested", time);
+    storeCard(cardEntity);
+  }
+
+  /**
+   * Takes a list of cards that have been sorted and breaks them into smaller lists that represent
+   * rounds.
+   */
+  private List<List<QuizCard>> createQuizRounds(List<Card> userCards) {
+    List<List<QuizCard>> quiz = new ArrayList<List<QuizCard>>();
+    int numRounds = userCards.size() / numOfCardsPerRound;
+    int start = 0;
+    if (numRounds > maxNumOfRounds) {
+      numRounds = maxNumOfRounds;
+    }
+
+    for (int i = 0; i < numRounds; i++) {
+      quiz.add(new ArrayList<QuizCard>());
+      for (int j = start; j < (start + numOfCardsPerRound); j++) {
+        // Calls constructor that fetches similar words for the quiz
+        QuizCard quizCard = new QuizCard(userCards.get(j));
+        quiz.get(i).add(quizCard);
+      }
+      start += numOfCardsPerRound;
+    }
+
+    return quiz;
+  }
+
+  private Entity initializeCard(Entity entity) {
+    // If cards don't have a familiarity score then provide default value
+    if (!entity.hasProperty("familiarityScore")) {
+      entity.setProperty("familiarityScore", .5);
+      entity.setProperty("timeTested", System.currentTimeMillis());
+    }
+
+    return entity;
+  }
+
+  private Card storeCard(Entity entity) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Transaction txn = datastore.beginTransaction(Builder.withXG(true));
+    try {
+      datastore.put(txn, entity);
+      txn.commit();
+    } finally {
+      if (txn.isActive()) {
+        txn.rollback();
+      }
+    }
+    Card card = new Card(entity);
+    return card;
+  }
+
+  /** Parses optional parameters that allow for testing on fewer rounds with fewer cards. */
+  private void setOptionalQuizParameters(HttpServletRequest request) {
+    String optNumCards = request.getParameter("numCards");
+    String optNumRounds = request.getParameter("numRounds");
+    if (optNumCards != null && optNumRounds != null) {
+      try {
+        numOfCardsPerRound = Integer.parseInt(optNumCards);
+        maxNumOfRounds = Integer.parseInt(optNumRounds);
+      } catch (NumberFormatException e) {
+        numOfCardsPerRound = 5;
+        maxNumOfRounds = 4;
+      }
+    }
+  }
+
```

Comment:
This is sketchy — now you're persisting extra state in the servlet unnecessarily. Better to have this function return `numOfCardsPerRound` and `maxNumOfRounds` and not modify servlet state.

I also don't see this function used anywhere.

Reply:
This function was meant only for my tests. So that they don't need over 5 cards to run a test.

Reply:
done.